### PR TITLE
Instance endpoints

### PR DESCRIPTION
**Pull request by kaihoffman**
Civo instance endpoints finished - 
List
Show
Create
Tags
Update (hostname/notes)
Remove
Reboot
Soft-Reboot
Console
Stop
Start
Upgrade (size)
Move IP
Firewall (set instance to use)

### CODE REVIEW

**Comment by kaihoffman** [comment]
Partial name matching for all instance methods should work now. (So you can search by partial UUID or hostname, and it returns a friendly error if the result is 0 or >1).

I also have some other endpoints (firewall, config) in another branch.

File: lib/instance.rb [hunk 1]
```diff
@@ -79,65 +80,192 @@ def show(id)
       exit 1
     end
 
-    desc "", ""
-    def create
+    desc "create --hostname=host_name --size=instance_size [--template= ][--snapshot= ]", "create a new instance with specified hostname, instance size, template/snapshot ID. Optional: region, public_ip (true or false), initial user"
```
Comment (andyjeffries) [inline lib/instance.rb]: Feels like the help text is non-standard (and there are others like this, so please review them all):

`--size=instance_size [--template= ][--snapshot= ]`

Should be more like:

`--size=instance_size [--template=template_id] [--snapshot=snapshot_id]`

e.g. a space in between options and an example value showing clearly it's the ID (not the name).

File: lib/instance.rb [hunk 2]
```diff
@@ -79,65 +80,192 @@ def show(id)
...
+    option :hostname, :required => true
```
Comment (andyjeffries) [inline lib/instance.rb]: These should all be in the "new" hash syntax, e.g.:

`option :hostname, required: true`

File: lib/instance.rb [hunk 3]
```diff
@@ -79,65 +80,192 @@ def show(id)
...
+        puts "Please provide either template OR snapshot ID".colorize(:red)
+        exit 1
+      end
+      
+      if options[:template]
+        Civo::Instance.create(hostname: options[:hostname], size: options[:size], template: options[:template], initial_user: options[:initial_user], region: options[:region], ssh_key_id: options[:ssh_key_id], tags: options[:tags])
+      end
+
+      if options[:snapshot]
+      Civo::Instance.create(hostname: options[:hostname], size: options[:size], snapshot_id: options[:snapshot], initial_user: options[:initial_user], region: options[:region], ssh_key_id: options[:ssh_key_id], tags: options[:tags])
```
Comment (andyjeffries) [inline lib/instance.rb]: Indentation isn't correct.

File: lib/instance.rb [hunk 4]
```diff
@@ -79,65 +80,192 @@ def show(id)
...
+      end
+      
+      puts "        Created instance #{options[:hostname].colorize(:green)}"
```
Comment (andyjeffries) [inline lib/instance.rb]: Why does this have a load of blank spaces before "Created"?

File: lib/instance.rb [hunk 5]
```diff
@@ -79,65 +80,192 @@ def show(id)
...
 
-    desc "tags ID", ""
-    def tags
+    desc "tags ID 'tag1 tag2 tag3...'", "retag instance by ID (input no tags to clear all tags)"
+    def tags(id, newtags=nil)
+      CivoCLI::Config.set_api_auth
+      instance = detect_instance_id(id)
 
+        Civo::Instance.tags(id: instance.id, tags: newtags)
+        puts "        Updated tags on #{instance.hostname.colorize(:green)}. Use 'civo instance show #{instance.hostname}' to see the current tags.'"
```
Comment (andyjeffries) [inline lib/instance.rb]: Again, weird leading blank space.

File: lib/instance.rb [hunk 6]
```diff
@@ -79,65 +80,192 @@ def show(id)
...
+      rescue Flexirest::HTTPException => e
+      puts e.result.reason.colorize(:red)
```
Comment (andyjeffries) [inline lib/instance.rb:122]: Indentation is not correct.

File: lib/instance.rb [hunk 7]
```diff
@@ -79,65 +80,192 @@ def show(id)
...
+      exit 1
     end
 
-    desc "", ""
-    def update
+    desc "update ID/HOSTNAME [--name new_hostname][--notes 'txt']", "update details of instance. Use --hostname=new_name, --notes='notes' to specify update"
```
Comment (andyjeffries) [inline lib/instance.rb]: As above, spacing isn't correct. Also, these don't use `=` in between option keys and values (but the ones above did?)

File: lib/instance.rb [hunk 8]
```diff
@@ -79,65 +80,192 @@ def show(id)
...
+
+     instance = detect_instance_id(id)
+      
+      puts "        Rebooting #{instance.hostname.colorize(:red)}. Use 'civo instance show #{instance.hostname}' to see the current status."
+      instance.reboot
 
-    desc "", ""
-    def hard_reboot
-      # {ENV["CIVO_API_VERSION"] || "1"}/instances/:id/hard_reboots", requires: [:id]
+      rescue Flexirest::HTTPException => e
```
Comment (andyjeffries) [inline lib/instance.rb]: Indentation isn't correct.

File: lib/instance.rb [hunk 9]
```diff
@@ -150,11 +278,40 @@ def move_ip
...
+      end
+      result.select! {|instance| instance.hostname.include?(id) }
+      
+      if result.count == 0
+        puts "No instances found for '#{id}'. Please check your query."
+        exit 1
+      elsif result.count > 1
+        puts "Multiple possible instances found for '#{id}'. Please try with a more specific query."
+        exit 1
+      else result[0]
```
Comment (andyjeffries) [inline lib/instance.rb]: The `result[0]` should be on a separate line to the `else`, this looks like a mistake.

**Review by andyjeffries — CHANGES_REQUESTED** [review]
I've given examples of where things need fixing, but haven't necessarily noted all instances, e.g.:

1) Help text needs spaces between options, consistent use of `=`
2) Indentation issues, particularly around the `rescue` construct
3) Spacing before success/failure output

File: lib/instance.rb (hunk 2 above)
Comment (kaihoffman) [reply lib/instance.rb]: All edited.

File: lib/instance.rb (hunk 1 above)
Comment (kaihoffman) [reply lib/instance.rb]: Edited all help texts to unify syntax and messaging.

File: lib/instance.rb (hunk 4 above)
Comment (kaihoffman) [reply lib/instance.rb]: Edited (the original intention was to match the output of another command but to be honest I have forgotten which one it was)

File: lib/instance.rb (hunk 6 above)
Comment (kaihoffman) [reply lib/instance.rb:122]: fixed

File: lib/instance.rb (hunk 7 above)
Comment (kaihoffman) [reply lib/instance.rb]: Spacing and equals signs added.

File: lib/instance.rb (hunk 8 above)
Comment (kaihoffman) [reply lib/instance.rb]: Fixed

File: lib/instance.rb (hunk 2 above)
Comment (kaihoffman) [reply lib/instance.rb]: Changed.

File: lib/instance.rb (hunk 1 above)
Comment (kaihoffman) [reply lib/instance.rb]: standardised

**Comment by kaihoffman** [comment]
Edits made as per request above: https://github.com/civo/cli/pull/9/commits/1293b952d07ed1bbe54f7026c9566ad16e35e080

**Pull request merged**